src/object_store/src/object/s3.rs

```diff
@@ -62,7 +63,10 @@ impl ObjectStore for S3ObjectStore {
         };
 
         let resp = req.send().await?;
-        let val = resp.body.collect().await?.into_bytes();
+
+        let mut body = resp.body.collect().await?;
+        let mut val = vec![0; body.remaining()];
+        body.copy_to_slice(&mut val);
```

Comment:
I think we can use something like `into` to avoid copy

Reply:
It's impossible here. The `.collect()` returns `aws-http-smith::byte_stream::AggregatedBytes`, which can be  discontinuous:

```rust
/// Non-contiguous Binary Data Storage
///
/// When data is read from the network, it is read in a sequence of chunks that are not in
/// contiguous memory. [`AggregatedBytes`](crate::byte_stream::AggregatedBytes) provides a view of
/// this data via [`impl Buf`](bytes::Buf) or it can be copied into contiguous storage with
/// [`.into_bytes()`](crate::byte_stream::AggregatedBytes::into_bytes).
#[derive(Debug, Clone)]
pub struct AggregatedBytes(SegmentedBuf<Bytes>);

impl AggregatedBytes {
    /// Convert this buffer into [`Bytes`](bytes::Bytes)
    ///
    /// # Why does this consume `self`?
    /// Technically, [`copy_to_bytes`](bytes::Buf::copy_to_bytes) can be called without ownership of self. However, since this
    /// mutates the underlying buffer such that no data is remaining, it is more misuse resistant to
    /// prevent the caller from attempting to reread the buffer.
    ///
    /// If the caller only holds a mutable reference, they may use [`copy_to_bytes`](bytes::Buf::copy_to_bytes)
    /// directly on `AggregatedBytes`.
    pub fn into_bytes(mut self) -> Bytes {
        self.0.copy_to_bytes(self.0.remaining())
    }
}
```

Its `into_bytes()` implementation also called `SegmentedBuf::copy_to_bytes`, which can avoid copy only when all remaining bytes are in its first segment.